core/src/test/java/org/apache/iceberg/rest/TestHTTPClient.java

```diff
@@ -0,0 +1,250 @@
+      mockRequest = mockRequest.withBody(asJson);
+    }
+
+    HttpResponse mockResponse = response().withStatusCode(200);
+    if (method.usesResponseBody()) {
+      mockResponse = mockResponse.withBody(asJson);
+    }
+
+    mockServer
+        .when(mockRequest)
+        .respond(mockResponse);
+
+    Item successResponse = null;
+    switch (method) {
+      case POST:
+        successResponse = restClient.post(path, body, Item.class, onError);
+        break;
+      case GET:
+        successResponse = restClient.get(path, Item.class, onError);
+        break;
+      case HEAD:
+        restClient.head(path, onError);
+        break;
+      case DELETE:
+        successResponse = restClient.delete(path, Item.class, onError);
+        break;
+    }
+
+    if (method.usesRequestBody()) {
+      Assert.assertEquals("On a successful " + method + ", the correct response body should be returned",
+          successResponse, body);
+    }
+    Assert.assertEquals("On a successful " + method + ", the error handler should not be called",
+        0, errorCounter.get());
+  }
+
+  public static void testHttpMethodOnFailure(HttpMethod method) throws JsonProcessingException {
+    Item body = new Item(0L, "hank");
+    int statusCode = 404;
+    String asJson = MAPPER.writeValueAsString(body);
+    AtomicInteger errorCounter = new AtomicInteger(0);
+    ErrorResponse response = ErrorResponse.builder()
+        .responseCode(statusCode).withMessage("Not found").build();
+    Consumer<ErrorResponse> onError = error -> {
+      System.out.println("In the error handler for " + method + " and received " + ErrorResponseParser.toJson(error));
+      errorCounter.incrementAndGet();
+      throw new RuntimeException(
+          String.format("Called error handler for method %s due to status code: %d", method, statusCode));
+    };
+
+    // Using different paths keeps the expectations unique for the test's mock server
+    String path = String.format("%s_failure", method);
+    HttpRequest mockRequest = request("/" + path).withMethod(method.name().toUpperCase(Locale.ROOT));
+    if (method.usesRequestBody()) {
+      mockRequest = mockRequest.withBody(asJson);
+    }
+
+    HttpResponse mockResponse = response().withStatusCode(statusCode);
+    if (method.usesResponseBody()) {
+      mockResponse = mockResponse.withBody(ErrorResponseParser.toJson(response));
+    }
+
+    mockServer
+        .when(mockRequest)
+        .respond(mockResponse);
+
+    Assertions
+        .assertThrows(RuntimeException.class, () -> {
+          switch (method) {
+            case POST:
+              restClient.post(path, body, Item.class, onError);
+              break;
+            case GET:
+              restClient.get(path, Item.class, onError);
+              break;
+            case HEAD:
+              restClient.head(path, onError);
+              break;
+            case DELETE:
+              restClient.delete(path, Item.class, onError);
+              break;
```

Comment:
Minor: I think this could be deduped between this and the success case with a method that runs the request and either succeeds and returns `Item` (null for head) or fails.

Reply:
Made this into its own function that is shared between the success and failure case.